repos/system_upgrade/common/actors/selinux/selinuxapplycustom/actor.py

```diff
@@ -35,15 +39,42 @@ def process(self):
             self.log.warning("Failed to create working directory! Aborting.")
             return
 
+        # get list of policy modules after the upgrade
+        installed_modules = set(
+            [x for (x, _) in selinuxapplycustom.list_selinux_modules()]
+        )
+
         # import custom SElinux modules
         for semodules in self.consume(SELinuxModules):
-            self.log.info("Processing custom SELinux policy modules. Count: %d.", len(semodules.modules))
+            self.log.info(
+                "Processing custom SELinux policy modules. Count: %d.",
+                len(semodules.modules),
+            )
             for module in semodules.modules:
-                cil_filename = os.path.join(WORKING_DIRECTORY, "{}.cil".format(module.name))
-                self.log.info("Installing module %s on priority %d.", module.name, module.priority)
+                # check for presence of udica templates and make sure to install their latest versions
+                selinuxapplycustom.install_udica_templates(semodules.templates)
+                # Skip modules that are already installed. This prevents DSP modules installed with wrong
+                # priority (usually 400) from being overwritten by an older version
+                if module.name in installed_modules:
+                    self.log.info(
+                        "Skipping module %s on priority %d because it is already installed.",
+                        module.name,
+                        module.priority,
+                    )
+                    continue
+
+                # cil module files need to be extracted to disk in order to be installed
+                cil_filename = os.path.join(
+                    WORKING_DIRECTORY, "{}.cil".format(module.name)
+                )
+                self.log.info(
+                    "Installing module %s on priority %d.", module.name, module.priority
+                )
                 if module.removed:
-                    self.log.warning("The following lines where removed because of incompatibility: \n%s",
-                                     '\n'.join(module.removed))
+                    self.log.warning(
+                        "The following lines where removed because of incompatibility: \n%s",
```

Comment:
Please use format function for string formatting, we have it in guidelines https://leapp.readthedocs.io/en/latest/python-coding-guidelines.html#prefer-new-style-string-formatting-to-old-style
(Won't mark other occurrencies not to be too annoying)